Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp

```diff
@@ -0,0 +1,264 @@
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QListWidget>
+#include <QPlainTextEdit>
+#include <QPushButton>
+#include <QSlider>
+#include <QSpinBox>
+#include <QVBoxLayout>
+
+static constexpr int SLIDER_TICK_COUNT = 100;
+
+IOWindow::IOWindow(MappingButton* parent, ControlReference* ref, IOWindow::Type type)
+    : QDialog(static_cast<QWidget*>(parent)), m_reference(ref), m_parent(parent), m_type(type)
+{
+  CreateMainLayout();
+  ConnectWidgets();
+  setWindowTitle(type == IOWindow::Type::Input ? tr("Configure Input") : tr("Configure Output"));
+}
+
+void IOWindow::CreateMainLayout()
+{
+  m_main_layout = new QVBoxLayout();
+
+  m_devices_combo = new QComboBox();
+  m_option_list = new QListWidget();
+  m_select_button = new QPushButton(tr("Select"));
+  m_detect_button = new QPushButton(tr("Detect"));
+  m_or_button = new QPushButton(tr("| OR"));
+  m_and_button = new QPushButton(tr("&& AND"));
+  m_add_button = new QPushButton(tr("+ ADD"));
+  m_not_button = new QPushButton(tr("! NOT"));
+  m_test_button = new QPushButton(tr("Test"));
+  m_expression_text = new QPlainTextEdit();
+  m_button_box = new QDialogButtonBox();
+  m_clear_button = new QPushButton(tr("Clear"));
+  m_apply_button = new QPushButton(tr("Apply"));
+  m_range_slider = new QSlider(Qt::Horizontal);
+  m_range_spinbox = new QSpinBox();
+
+  // Devices
+  m_main_layout->addWidget(m_devices_combo);
+
+  // Range
+  auto* range_hbox = new QHBoxLayout();
+  range_hbox->addWidget(new QLabel(tr("Range")));
+  range_hbox->addWidget(m_range_slider);
+  range_hbox->addWidget(m_range_spinbox);
+  m_range_slider->setMinimum(-500);
+  m_range_slider->setMaximum(500);
+  m_range_spinbox->setMinimum(-500);
+  m_range_spinbox->setMaximum(500);
+  m_main_layout->addItem(range_hbox);
+
+  // Options (Buttons, Outputs) and action buttons
+  for (QPushButton* button : {m_select_button, m_detect_button, m_or_button, m_and_button,
+                              m_add_button, m_not_button, m_test_button})
+    button->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Expanding);
+
+  auto* hbox = new QHBoxLayout();
+  auto* button_vbox = new QVBoxLayout();
+  hbox->addWidget(m_option_list);
+  hbox->addItem(button_vbox);
+
+  button_vbox->addWidget(m_select_button);
+  button_vbox->addWidget(m_type == Type::Input ? m_detect_button : m_test_button);
+  button_vbox->addWidget(m_or_button);
+
+  if (m_type == Type::Input)
+  {
+    button_vbox->addWidget(m_and_button);
+    button_vbox->addWidget(m_add_button);
+    button_vbox->addWidget(m_not_button);
+  }
+
+  m_main_layout->addItem(hbox);
+  m_main_layout->addWidget(m_expression_text);
+
+  // Button Box
+  m_main_layout->addWidget(m_button_box);
+  m_button_box->addButton(m_clear_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(m_apply_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(QDialogButtonBox::Ok);
+
+  setLayout(m_main_layout);
+}
+
+void IOWindow::Update()
+{
+  if (m_controller == nullptr)
```

Comment:
How does this work? Isn't the parent window disabled while this window is visible?